### ISSUE TITLE
PragmataPro Mono rendered as some kind of german font

### ISSUE THREAD

**Issue by rubin55** [neovide/neovide#2786]
**Describe the bug**
Set PragmataPro Mono font in config.toml, observe strange german font being used.

**To Reproduce**

config.toml:
```toml
tabs = false

[font]
normal = ["PragmataPro Mono"]
size = 13.0
hinting = "slight"
edging = "antialias"
```

1. Install PragmataPro font
2. Use above config file
3. Observe as seen in screenshot

**Expected behavior**
Neovide should use the specified font and not this strange gothic german one.

**Screenshots**
![image](https://github.com/user-attachments/assets/0fa388ac-48e0-4881-b882-a2130525b757)

**Desktop (please complete the following information):**
 - OS: Arch Linux
 - Neovide Version 0.13.3
 - Neovim Version 0.10.1

**Please run `neovide --log` and paste the contents of the `.log` file created in the current directory here:**
[neovide_rCURRENT.log](https://github.com/user-attachments/files/16961909/neovide_rCURRENT.log)

**Additional context**
N/A.

**Comment by rubin55** [comment]
Thanks, I tried that, but I keep getting this german font. It's really weird. Is there some command that you can use from within neovide that lists the fonts it "sees"?

**Comment by fredizzimo** [comment]
Yes `:set guifont=*`. But I think the font name is already understood, since otherwise you would get an error. The log files also shows that a font with the correct name is loaded.

**Comment by fredizzimo** [comment]
Ok, I managed to match the font, it looks like it's loading `pragmatapro-fraktur`
https://fonts.adobe.com/fonts/pragmatapro-fraktur.

So, yes, check the output of `set guifont=*` to see if it's listed by another name than `PragmataPro Mono`

**Comment by rubin55** [comment]
Yeah that's exactly what it looks like, however, I only have PragmataPro, PragmataPro Liga, PragmataPro Mono and PragmataPro Mono Liga installed, not that font, so my gut feeling is that it's embedded somehow; I've asked the font creator here: https://github.com/fabrizioschiavi/pragmatapro/issues/300

**Comment by fredizzimo** [comment]
I took another look, it actually looks like fraktur is enabled as `ss03`, see
https://github.com/fabrizioschiavi/pragmatapro/blob/master/useful_files/Handbook.png

But I wonder why it gets enabled, you could try disabling it
```toml
[font]
normal = ["PragmataPro Mono"]
size = 13.0
hinting = "slight"
edging = "antialias"

[font.features]
"PragmataPro Mono"" = [ "-ss03" ]
```

I would also assume that it only applies to the italic style. But in your screenshot the regular font style seem to be affected as well.

**Comment by rubin55** [comment]
Hi @fredizzimo, here is the output of `fc-match -v "PragmataPro Mono"`:

```shell
# fc-match -v "PragmataPro Mono"
Pattern has 44 elts (size 48)
	family: "PragmataPro Mono"(s)
	familylang: "en"(s)
	style: "Regular"(s)
	stylelang: "en"(s)
	fullname: "PragmataPro Mono Regular"(w)
	fullnamelang: "en"(s)
	slant: 0(i)(s)
	weight: 80(f)(s)
	width: 100(f)(s)
	size: 12(f)(s)
	pixelsize: 12.5(f)(s)
	spacing: 100(i)(s)
	foundry: "FSD "(s)
	antialias: True(w)
	hintstyle: 1(i)(w)
	hinting: True(w)
	verticallayout: False(s)
	autohint: True(w)
	globaladvance: True(s)
	file: "/usr/share/fonts/pragmatapro/PragmataPro_Mono_R_0830.ttf"(s)
	index: 0(i)(w)
	outline: True(s)
	scalable: True(s)
	dpi: 192(f)(w)
	rgba: 5(i)(w)
	scale: 1(f)(s)
	charset:
	0000: 08000000 ffffffff ffffffff 7fffffff 00000000 ffffffff ffffffff ffffffff
	0001: ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff
	0002: ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff
	0003: ffffffff ffffffff ffffffff 4000ffff ffffd7f0 fffffffb 307f7fff 00330000
	0004: ffffffff ffffffff ffffffff 000c803f 0f030000 00000000 00000000 00000000
	0005: 00000000 00000000 00000000 00000000 fffe0000 ffffffff ffff00ff 001f07ff
	0006: c82036c0 07fffffe 7fffffff ffffffff ffffffff ffffffff 9fffffff fffffdff
	0007: 00000000 00000000 ffff0000 00003fff 00000000 00000000 00000000 00000000
	0016: 00000000 00000000 00000000 00000000 00000000 ffffffff ffffffff 01ffffff
	001a: 00000000 00000000 00000000 00000000 00000000 ffff0000 00000001 00000000
	001d: ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff 3400ffff fbe24951
	001e: ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff
	001f: 3f3fffff ffffffff aaff3f3f 3fffffff ffffffff 1fdfffff efcfffde 3fdcffff
	0020: ffffffff ffffffff ffffffff fff3ffff 1fff7fff a62e1618 ffff8000 0001ffff
	0021: ffffffff ffffffff ffffffff ffffffff ffff0fff ffffffff ffffffff ffffffff
	0022: ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff
	0023: ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff
	0024: ffffffff 0000007f 00000000 000001ff 00000000 ffc00000 ffffffff 000007ff
	0025: ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff
	0026: fdffffff feffe7ff ffffffff 07f8feff 0207ffff 3ffffdff 00100000 00000000
	0027: 05f8010e 00002000 07001000 ffffff3c ffffffff ffffffff ffffffff ffffffff
	0028: ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff
	0029: 1e0c0fc0 01c00000 fffffe7f ffffffff 179fffff f8000000 5fffcc00 fffffff8
	002a: 9fffe007 f03ffffe fffef3ff fe8fffff ffffffff e7ffffdf 7e7f9fff fe7ff000
	002b: ffffefef 0002cfe0 ffc7e000 800affff ffe0f5ff 2100ffff 0000f1ef 8000a000
	002c: 00000000 00000000 00000000 30020000 00000000 00000000 00000000 00000000
	002e: 00000000 00004000 00000000 00000000 00000000 00000000 00000000 00000000
	0031: ffffffe0 000003ff 00000000 00000000 00000000 00000000 00000000 00000000
	00ab: 00000000 00000000 00000000 00000c00 00000000 00000000 00000000 00000000
	00e0: 0001ffff 00000000 00000000 00000000 00000000 ffffc00f 0017f5ff 00000000
	00e1: 00000000 00000000 00000000 00000000 00000000 00000000 0000003f 00000000
	00e2: ffffffff ffffffff ffffffff ffffffff ffffffff 007f03ff 00007fff 7fff7fff
	00e3: ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff 0000000f
	00e4: 00000000 00000000 00000000 00000000 00000000 00000000 00001fff 00000000
	00e5: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 fc000000
	00e6: ffffffff ffffffff ffffffff ffffffff ffffffff 0007ffff 00000000 00000000
	00e7: ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff 0000003f 00000000
	00e9: 00000000 04000000 00003041 00000000 00000000 00000000 00000000 00000000
	00ea: 00000000 00000000 00000000 ffffffff ffff9dff ffffffff fffff2ff ffffffff
	00eb: fffffbff ffffffff ffff7fff ffffffff ffffffff ffffffff ffffffff 00000fff
	00ee: 00000fff 00000000 00000000 00000000 00000000 00000000 00000000 00000000
	00f0: 7fff7fff 7ffffffe 7fff7fff 7fff7fff 7fff7fff 00077fff 7fff7fff 7fff7fff
	00f1: 7fff7fff 7fff7fff 7fff7fff 7fff7fff 7fff7fff 7fff7fff 7fff7fff 7fff7fff
	00f2: 7fff7fff 7ffffffe 7fff7fff 7fff7fff 7fff7fff 7fff7fff 7fff7fff 00000001
	00f3: ffffffff ffffffff ffffffff 0007ffff 00000000 00000000 00000000 00000000
	00f4: ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff
	00f5: ffffffff 000fffff 00000000 00000000 00000000 00000000 00000000 00000000
	00fb: e0000007 5f7fffff ffffffdb ffffffff ffffffff fffffcff fff80003 f00003ff
	00fc: 00000000 00000000 f8000000 0000000f 0001c000 00000000 00000000 00000000
	00fd: 00000000 30000000 00000000 00000000 00000000 00000000 00000000 00000000
	00fe: 00000000 800f3fff fff70187 ffff0f7f ffffffff ffffffff ffffffff 1fffffff
	00ff: fffffffe ffffffff 7fffffff 0000003e 00000000 00000000 00000000 3fff7f7f
	01d4: ffffffff ffffffff ffffffff ffffffff ffffffff efffffff ffffffff ffffffff
	01d5: ffffffff 7bffffff fffdfc5f ffffffff ffffffff ffffffff ffffffff ffffffff
	01d6: ffffffff ffffffff ffffffff ffffffff ffffffff ffffff3f ffffffff ffffffff
	01d7: ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffcfff ffffffff
	01f0: 00000000 00000000 00000000 00000000 00000000 7ffe7fff fffefffe 00000000
	01f1: 0000e000 00008000 00000000 0000e000 00000000 00000000 00000000 00000000
	01f8: 00000000 00000000 00000000 0000001f 00000000 00000000 00000000 00000000
	01fb: ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff 000007ff 03ff0600
	0f00: fffffffe ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff
	0f01: ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff
	0f02: ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff
	0f03: ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff
	0f04: ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff
	0f05: ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff
	0f06: ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff
	0f07: ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff
	0f08: ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff
	0f09: ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff
	0f0a: ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff
	0f0b: ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff
	0f0c: ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff
	0f0d: ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff
	0f0e: ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff
	0f0f: ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff
	0f10: ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff
	0f11: ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff
	0f12: ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff
	0f13: ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff
	0f14: ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff
	0f15: ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff
	0f16: ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff
	0f17: ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff
	0f18: ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff
	0f19: ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff
	0f1a: ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff ffffffff 0001ffff
	1000: 00010000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
	1001: 0000000f 00000000 0007f01f 0007f001 0001c000 0201ffff 3fff0000 007ff000
	1002: ffffffff 0003ffff 000fffff 8007ffff 3fff0318 fffffe00 0081bc00 00f80001
	1003: 00000000 00000000 00000fff 07fffff8 ffffffff 000fffff fffffe7f ffffffff
	1004: 0000003f ffffffff fffffff0 f0ffffff 000001ff 00000000 00000000 0000ffff
	1005: 07ffffff ffffffff 000fffff fffffe7f ffffffff 0000003f ffffffff fffffff0
	1006: f0ffffff 000001ff 00000000 00000000 00000000 00000000 00000000 00000000
(s)
	lang: aa|af|ar|ast|av|ay|az-az|az-ir|bm|be|bg|bi|bin|br|bs|ca|ce|ch|co|cs|cy|da|de|el|en|eo|es|et|eu|fa|fi|fj|fo|fr|ff|fur|fy|ga|gd|gl|gn|gv|ha|haw|he|ho|hr|hu|ia|ig|id|ie|ik|io|is|it|ki|kl|ku-ir|kum|kw|la|lb|lez|ln|lt|lv|mg|mh|mi|mk|mo|mt|nb|nds|nl|nn|no|nr|nso|ny|oc|om|os|pl|ps-af|ps-pk|pt|rm|ro|ru|sco|se|sel|shs|sk|sl|sm|sma|smj|smn|sms|so|sq|sr|ss|st|sv|sw|tk|tl|tn|to|tr|ts|tw|ug|uk|ur|uz|ve|vi|vo|vot|wa|wen|wo|xh|yap|yi|yo|zu|ak|an|ber-dz|crh|csb|ee|fat|fil|hsb|ht|hz|jv|kab|kj|kr|ku-iq|ku-tr|kwm|lg|li|ms|na|ng|nv|ota|pa-pk|pap-an|pap-aw|qu|quz|rn|rw|sc|sd|sg|sn|su|ty|za|lah|und-zmth|agr|ayc|bem|ckb|dsb|lij|mfe|mhr|miq|mjw|nhn|niu|rif|sgs|szl|tpi|unm|wae|yuw(s)
	fontversion: 54394(i)(s)
	capability: "otlayout:DFLT otlayout:arab otlayout:bopo otlayout:brai otlayout:cyrl otlayout:grek otlayout:hebr otlayout:latn otlayout:math otlayout:zinh otlayout:zyyy"(w)
	fontformat: "TrueType"(s)
	embeddedbitmap: True(s)
	decorative: False(s)
	lcdfilter: 0(i)(w)
	namelang: "en"(s)
	prgname: "fc-match"(s)
	postscriptname: "PragmataProMono-Regular"(s)
	color: False(s)
	symbol: False(s)
	variable: False(s)
	fonthashint: True(s)
	order: 0(i)(s)
	desktop: "GNOME"(s)
	namedinstance: False(s)
```

**Comment by fredizzimo** [comment]
I don't see anything out of the ordinary, except that it seems to be a ttf font, which if I understand correctly has a very limited support for font features, so it's possible that the Skia library we use, just enables all of them. According to this otf files should be available though https://github.com/fabrizioschiavi/pragmatapro/issues/104

You could try dropping the ttf https://fontdrop.info/, to see what's enabled by default.

BTW, did you see this message https://github.com/neovide/neovide/issues/2786#issuecomment-2356351770, and try the config to disable the features?

**Comment by fredizzimo** [comment]
Actually, it's https://github.com/dfrg/swash that makes use of the feature information, which makes it more likely that they are handled wrong for ttf fonts.

**Comment by rubin55** [comment]
I did check the font with fontdrop.info. It can see 32 opentype styles, only `calt` seems to be enabled by default. Do you think this is a bug in swash?

**Comment by fredizzimo** [comment]
Maybe, especially if it can't be disabled by
```toml
[font.features]
"PragmataPro Mono" = [ "-ss03" ]
```

If they provide an OTF font, like https://github.com/fabrizioschiavi/pragmatapro/issues/104 suggests, you could try that.

You could also try to generate a custom font with the feature removed using https://github.com/twardoch/fonttools-opentype-feature-freezer, but that's only a hack.

Unfortunately, I'm not able to help debugging too much without buying the font, which I might have to do, but for now I'm still waiting for response from @fabrizioschiavi, in the issue you created.

**Comment by fredizzimo** [comment]
@rubin55, based on the response here https://github.com/fabrizioschiavi/pragmatapro/issues/300#issuecomment-2358008982, can you try disabling aalt?

```
[font.features]
"PragmataPro Mono" = [ "-aalt" ]
```

**Comment by rubin55** [comment]
That seems to result in a normal PragmataPro look:
![Screenshot from 2024-09-18 15-31-55](https://github.com/user-attachments/assets/206179a6-c87b-4deb-905c-767ae11df325)

**Comment by fredizzimo** [comment]
Thanks. Based on this information I will open an issue in the Swash repository a bit later.

**Comment by fredizzimo** [comment]
I reported the issue here:
* https://github.com/dfrg/swash/issues/62

**Comment by mrschyte** [comment]
I'm using ArchLinux and I've downgraded the neovide package from (0.13.3-3 => 0.13.3-2) and the issue is no longer present. Something must have changed between these two versions.

**Comment by fredizzimo** [comment]
@mrschyte, hm something does not add up there, 0.13.3-3 was released yesterday, and this issue is reported earlier. And I can't also not see how the changes could affect it.

https://gitlab.archlinux.org/archlinux/packaging/packages/neovide/-/commit/b0356f7265edb72c3c7797b5dad32a808595b115

Neovide 0.13.2, to 0.13.3 is more likely. But even if it touches the font features, it should only affect customizations and fallbacks https://github.com/neovide/neovide/compare/0.13.2...0.13.3

**Comment by mrschyte** [comment]
Yeah, I've checked the diff and I find it strange as well. However, I didn't have this issue before upgrading yesterday.

I've tested it again using the packages in /var/cache/pacman; with 0.13.3-3 the issue is present, with 0.13.3-2, the font displays fine. I'm currently rebuilding the neovide package from source and I'll check with that as well.

These are the sha1 checksums from the package cache:

`2b32619a029c4593a5c7a6ea47eb5913297eba08  pkg/neovide-0.13.3-2-x86_64.pkg.tar.zst
37a2eb1caafacc1fa8cb44d016be039f6a02bddc  pkg/neovide-0.13.3-2-x86_64.pkg.tar.zst.sig
20a48c7b280dd791d1cbfb0efd961fc0b16cec0b  pkg/neovide-0.13.3-3-x86_64.pkg.tar.zst
7d0fe89a3340017ad4e948a8960e463d764f9342  pkg/neovide-0.13.3-3-x86_64.pkg.tar.zst.sig`

**Comment by mrschyte** [comment]
I've rebuilt both package versions and the issue is present in both. This probably means that one of the dependencies have changed between the two package releases. The swash versions however are the same (v0.1.17). I'll try to figure out which library has changed.

**Comment by fredizzimo** [comment]
It builds it using `--locked`, so the rust dependencies will be the same https://gitlab.archlinux.org/archlinux/packaging/packages/neovide/-/blob/b0356f7265edb72c3c7797b5dad32a808595b115/PKGBUILD#L43

However the fontconfig library might be statically linked

**Comment by fredizzimo** [comment]
Yes, skia is built from source there, and that uses system libraries

**Comment by fredizzimo** [comment]
@rubin55. Did you run with the Arch Linux package or build from source?

**Comment by rubin55** [comment]
@fredizzimo When I reported the issue, I was running the `neovide` Arch linux package from extras, version 0.13.3-2 (so just `pacman -S neovide`, no self-built version)

**Comment by mrschyte** [comment]
I've rebuilt neovide with older system libs to no avail (I tried using the 2024/06/01 ArchLinux archive, since that's the oldest which have icu-75; previous versions won't build using the PKGBUILD).

I've also noticed that @rubin55 is using "PragmataPro Mono" vs "PragmataPro Mono Liga" which I have set. The ligature version works with 0.13.3-2, but breaks on 0.13.3-3. The non-ligature version doesn't work with either version and displays the fraktur-like font.

**Comment by yngwi** [comment]
I’m having the same problem with `-aalt`. However, I also can’t seem to get ligatures to work using any version of Pragmata Pro. I tried to enable different font features after disabling aalt like this:

```toml
[font.features]
"PragmataPro" = ["-aalt", "+calt", "+clig", "+liga"]
```

I’m not sure if this is correct, but it doesn’t change anything anyways. I don’t know anything about opentype and it’s features so I’m not sure if this should be working or not.

**Comment by fredizzimo** [comment]
I'm not sure if the ligatures are included in the standard font or not, but there's a separate `PragmataPro Liga` with ligatures mentioned in the readme https://github.com/fabrizioschiavi/pragmatapro

**Comment by yngwi** [comment]
Yes. It's the same with all variants.

**Comment by griffinwxk** [comment]
I also have the same issue intermittently (depending on neovide versions and PragmataPro versions). I'm on macos and build from source. For a while there was no problem (iirc for neovide <= 13.2 and >= 12.? and PragmataPro 0.829/0.830), but I just compiled 13.3 today and the issue reappeared.

In the past I've tested using [swash_demo](https://github.com/dfrg/swash_demo) and it seems indeed a swash problem, but that issue disappeared for me after I upgrade to the neovide/font versions mentioned above so I didn't open an issue.

**Comment by aerosol** [comment]
```
[font]
normal = []
size = 14.0

[font.features]
"PragmataPro Mono" = [ "-aalt" ]
"PragmataPro Mono Liga" = [ "-aalt" ]
```

Doesn't seem to work for me. Confirmed that `config.toml` is being read.

PragmataPro version 0.830
neovide 0.13.3

![image](https://github.com/user-attachments/assets/44ea92bf-ef6e-44b2-a78c-3d9ccf55a4a2)

Needless to say, this is hilarious.

**Comment by fredizzimo** [comment]
Since this can be repeated using the swash demo. Can you @griffinwxk, or someone else fill in some more details in the existing swash issue that I opened a while ago?
* https://github.com/dfrg/swash/issues/62

**Comment by griffinwxk** [comment]
> Since this can be repeated using the swash demo. Can you @griffinwxk, or someone else fill in some more details in the existing swash issue that I opened a while ago?
> 
>     * [The aalt feature seems to be enabled by default dfrg/swash#62](https://github.com/dfrg/swash/issues/62)

Sorry for the late reply. I just found some time to find the testing code I did before (and tested again; yes the issue still persists). I put it in the issue you created. As I mentioned in more detail there, the behavior in `swash_demo` does not completely match neovide.

I'm willing to help further testing. I'm not a developer myself and not very versed in github, so reply from me tends to a bit slower though.

**Comment by rlue** [comment]
Chiming in to say that disabling aalt does not fix the issue for me:

<img width="1072" alt="Screenshot 2024-11-27 at 10 28 52" src="https://github.com/user-attachments/assets/b2b65caf-9865-4418-92b3-819ab435fc29">

On top of that, regular ("essential") PragmataPro is also rendered with a strange combination of the Old English gothic and superscript sans-serif glyphs:

<img width="1072" alt="Screenshot 2024-11-27 at 10 30 40" src="https://github.com/user-attachments/assets/9f1517f1-fc28-45be-93ee-2c799fdda699">

**Comment by synic** [comment]
This is working for me:

```toml
[font]
normal = ["PragmataPro Mono Liga"]
size = 12.0

[font.features]
"PragmataPro Mono Liga" = [ "-aalt", "+calt", "+clig", "-ss03", "-ss04" ]
```

**Comment by fabrizioschiavi** [comment]
This configuration works too:

- create a file `init.lua` in `.config/nvim/`

- paste this code:
`vim.o.guifont = "PragmataPro Mono Liga:h13"`

**Comment by synic** [comment]
Hrmm, my configuration gets rid of the german look, but it doesn't display ligatures. Tried this as well:

```toml
[font]
normal = ["PragmataPro Mono"]
size = 12.0

[font.features]
"PragmataPro Mono" = [ "-aalt", "+calt", "+clig", "-ss03", "-ss04", "+liga" ]
```

@fabrizioschiavi that was definitely not working for me - I had to put the "--aalt" for it to work (but like I said, now the ligatures don't work)

**Comment by fabrizioschiavi** [comment]
I use Mac OS and Neovide 0.13.3 and this what I see when I type some ligatures using my configuration

<img width="436" alt="Image" src="https://github.com/user-attachments/assets/d7f4a978-3023-4018-834f-2a6271b21317" />

**Comment by fredizzimo** [comment]
@fabrizioschiavi, Is there any chance you could provide the font to me, so I can debug the issue without having to buy it?

I think there is some kind of random behavior based on the reports, so it might not always be reproducible, but if I can I will most likely be able to find the issue and send a PR to swash, or one of its dependent libraries. 

The bug does not seem to be os dependent, since the original issue is for arch linux. And there's at least one report for MacOS.

The original report was for Neovide  0.13.3, but the are reports for our newest version 0.14 as well. The current git version could be potentially have a fix, since swash was updated to 0.2

**Comment by fabrizioschiavi** [comment]
No problem @fredizzimo I'll send you the font files for testing it. Plase let me know your email address.

**Comment by fredizzimo** [comment]
@fabrizioschiavi, thanks, I sent an email to email you use in git.

**Comment by fredizzimo** [comment]
Thanks, I can reproduce the issue here, so hopefully I will be able to find out what's wrong.

**Comment by fredizzimo** [comment]
@fabrizioschiavi, @dfrg
I managed to find out something, but I don't know if it's a swash bug or font bug

Swash gets confused by the  `calt` feature being present in both `gsub` and `gpos`.  When it finds it in the `gpos` table it gets assigned index 0, which is the same as `aalt` in the `gsub` table, so it gets enabled instead of the requested `calt`.

The somewhat random behaviour can probably be explained by the sorting and binary search, in some cases it probably finds the correct `calt` and enables that.

I`m not deeply familiar with how the opentype is supposed to work, so one of you might have a better idea of how to proceed from here.

**Comment by fredizzimo** [comment]
It looks like the bug is on the Swash side, I found this in the specification https://learn.microsoft.com/en-us/typography/opentype/spec/chapter2#feature-table

> A feature table defines the implementation of a feature using one or more lookups. Feature tables defined within the GSUB table contain references to glyph substitution lookups; feature tables defined within the GPOS table contain references to glyph positioning lookups. *If a feature requires both glyph substitution and positioning actions, then Feature tables referenced using the same feature tag will need to be defined in both the GSUB and GPOS tables*.

**Comment by synic** [comment]
Just a note, I was also unable to successfully disable the "fraktur" (ss03, ss04) on [fontfreeze](https://mutsuntsai.github.io/fontfreeze/). You can disable "aalt" and ss03, ss04 and get a version of the font that _mostly_ works without messing around with font features in the `neovide.toml` file, however, I still haven't been able to get both ligatures enabled and fraktur disabled. Disabling the one seems to prevent the other from working.

@fabrizioschiavi what are the chances you can output a version of the font that just doesn't have frakture at all?

**Comment by fredizzimo** [comment]
@synic, I don't think this can be worked around by changing the font or freezing the features. Something will always be broken as long as calt is present in both `GPOS` and `GSUB`. 

I also just noticed another thing, even if the bug is fixed in Swash, I'm not sure if the `calt` feature gets properly enabled for the ligatures, at least not by default, since it's only set for 'GSUB' here https://github.com/dfrg/swash/blob/6adffd7876fe6d3425aab53353e99ee52e56e2fe/src/shape/at.rs#L246-L248

**Comment by kierun** [comment]
Chiming in as well… I have the same issue.

@synic 's solution seemed to work for me, but I managed to get it down to just `-aalt` like so:

```toml
[font.features]
"PragmataPro Mono Liga" = [ "-aalt"]
```

Here are some screenshots:

## Before

![Image](https://github.com/user-attachments/assets/4dff692b-bf94-42e7-a01a-f3d68b36123d)

## After 

![Image](https://github.com/user-attachments/assets/e98abf53-ed1b-4f92-8a1c-43a0e3158348)

This is the same thing I see if I use the AppImage v0.13.3. The AppImage for 0.14.0 is the same as the installation from `cargo install …`. 

Ligatures are broken with the above fix.

I am not sure how this helps, or not… I am 💯 happy to help debug this. Just let me know how I can help.

**Comment by dfrg** [comment]
> [@synic](https://github.com/synic), I don't think this can be worked around by changing the font or freezing the features. Something will always be broken as long as calt is present in both `GPOS` and `GSUB`.
> 
> I also just noticed another thing, even if the bug is fixed in Swash, I'm not sure if the `calt` feature gets properly enabled for the ligatures, at least not by default, since it's only set for 'GSUB' here https://github.com/dfrg/swash/blob/6adffd7876fe6d3425aab53353e99ee52e56e2fe/src/shape/at.rs#L246-L248

Since `calt` is a default enabled feature, we can add it to the position feature set as well after the bug is fixed.

**Comment by fredizzimo** [comment]
This has now been fixed in Swash, and here's a Neovide PR that takes it into use 
* https://github.com/neovide/neovide/pull/3025

You can download the artifacts/binaries from here if you don't want to compile yourself https://github.com/neovide/neovide/pull/3025/checks

**Comment by fabrizioschiavi** [comment]
Thanks @fredizzimo! What was the cause of this issue?

**Comment by fredizzimo** [comment]
@fabrizioschiavi 

The cause was identified here https://github.com/neovide/neovide/issues/2786#issuecomment-2646310419

And fixed in Swash here https://github.com/dfrg/swash/pull/84

Basically, it didn't deal correctly with CALT being present in both GPOS and GSUB.

**Comment by synic** [comment]
Looks like it's working for me! Thanks!

**Comment by kierun** [comment]
LGTM. Thank you!
